initialize/user_data.go

```diff
@@ -15,40 +15,407 @@
 package initialize
 
 import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
 	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
 	"log"
+	"mime"
+	"mime/multipart"
+	"net/mail"
+	"strings"
 
 	"github.com/flatcar/coreos-cloudinit/config"
+	"github.com/flatcar/coreos-cloudinit/system"
 )
 
-var (
-	ErrIgnitionConfig = errors.New("not a config (found Ignition)")
+type UserDataType string
+
+const (
+	CloudConfigType UserDataType = "cloud-config"
+	ScriptType      UserDataType = "script"
+	IgnitionType    UserDataType = "ignition"
+	UnknownType     UserDataType = "unknown"
 )
 
-func ParseUserData(contents string) (interface{}, error) {
-	if len(contents) == 0 {
-		return nil, nil
+func NewUserData(payload string, env *Environment) (*UserData, error) {
+	if len(payload) == 0 {
+		return &UserData{}, nil
 	}
 
-	switch {
-	case config.IsScript(contents):
-		log.Printf("Parsing user-data as script")
-		return config.NewScript(contents)
-	case config.IsCloudConfig(contents):
-		log.Printf("Parsing user-data as cloud-config")
-		cc, err := config.NewCloudConfig(contents)
+	if env == nil {
+		return nil, fmt.Errorf("environment is nil")
+	}
+
+	parts, err := partsFromUserData(payload, env)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing user-data: %w", err)
+	}
+	return &UserData{
+		Parts: parts,
+		env:   env,
+	}, nil
+}
+
+func multipartToUserDataParts(payload string, env *Environment) ([]UserDataPart, error) {
+	if env == nil {
+		return nil, fmt.Errorf("environment is nil")
+	}
+	reader := strings.NewReader(payload)
+	m, err := mail.ReadMessage(reader)
+	if err != nil {
+		return []UserDataPart{}, fmt.Errorf("error parsing multipart MIME: %w", err)
```

Comment:
Could this return `nil`? And same for other error cases in this function.
```suggestion
		return nil, fmt.Errorf("error parsing multipart MIME: %w", err)
```

Reply:
It could. Even though the value should be ignored if ```err``` is not nil, I have a nagging urge to use an empty array and avoid nil pointer de-reference if someone actually tries to access the value.